### PR TITLE
Change Null-safe equal operator from cross join to hash join

### PR DESCRIPTION
**Pull request by EmmyMiao87**
ISSUE-2136

This commit change the join method from cross join to hash join when the equal operator is Null-safe '<=>'.
It will improve the speed of query which has the Null-safe equal operator.
The finds_nulls field is used to save if there is Null-safe operator.
The finds_nulls[i] is true means that the i-th equal operator is Null-safe.
The equal function in hash table will return true, if both val and loc are NULL when finds_nulls[i] is true.

### CODE REVIEW

**Comment by EmmyMiao87** [comment]
#2155

**Comment by EmmyMiao87** [comment]
#2136

File: be/src/exec/hash_join_node.cpp [hunk 1]
```diff
@@ -286,6 +292,11 @@ Status HashJoinNode::open(RuntimeState* state) {
         _is_push_down = false;
     }
 
+    if (std::find(_is_null_safe_eq_join.begin(), _is_null_safe_eq_join.end(),
```
Comment (morningman) [inline be/src/exec/hash_join_node.cpp]: Add a comment here to explain why not push down

File: be/src/exec/hash_table.cpp [hunk 2]
```diff
@@ -42,12 +42,18 @@ const char* HashTable::_s_llvm_class_name = "class.doris::HashTable";
 
 HashTable::HashTable(const vector<ExprContext*>& build_expr_ctxs,
                      const vector<ExprContext*>& probe_expr_ctxs,
-                     int num_build_tuples, bool stores_nulls, int32_t initial_seed,
+                     int num_build_tuples, bool null_preserved, 
+                     const std::vector<bool> finds_nulls,
```
Comment (morningman) [inline be/src/exec/hash_table.cpp]:
```suggestion
                     const std::vector<bool>& finds_nulls,
```

File: be/src/exec/hash_table.h [hunk 3]
```diff
@@ -98,7 +98,9 @@ class HashTable {
     HashTable(
         const std::vector<ExprContext*>& build_exprs,
         const std::vector<ExprContext*>& probe_exprs,
-        int num_build_tuples, bool stores_nulls, int32_t initial_seed,
+        int num_build_tuples, bool stores_nulls, 
+        const std::vector<bool> finds_nulls,
```
Comment (morningman) [inline be/src/exec/hash_table.h]:
```suggestion
        const std::vector<bool>& finds_nulls,
```

File: be/src/exec/olap_scan_node.cpp [hunk 4]
```diff
@@ -194,8 +194,10 @@ Status OlapScanNode::open(RuntimeState* state) {
     RETURN_IF_CANCELLED(state);
     RETURN_IF_ERROR(ExecNode::open(state));
 
+    LOG(INFO) << "olap scan node conjunct in open " << Expr::debug_string(_conjunct_ctxs);
```
Comment (morningman) [inline be/src/exec/olap_scan_node.cpp]: should be VLOG

**Review by morningman — CHANGES_REQUESTED** [review]

File: be/src/exec/olap_scan_node.cpp (hunk 4 above)
Comment (EmmyMiao87) [reply be/src/exec/olap_scan_node.cpp]: I will remove it.

File: fe/src/main/java/org/apache/doris/planner/HashJoinNode.java [hunk 5]
```diff
@@ -53,8 +54,8 @@
 
     private final TableRef     innerRef;
     private final JoinOperator joinOp;
-    // conjuncts of the form "<lhs> = <rhs>", recorded as Pair(<lhs>, <rhs>)
-    private  List<Pair<Expr, Expr>> eqJoinConjuncts;
+    // predicates of the form "<lhs> = <rhs>", recorded as Pair(<lhs>, <rhs>)
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/planner/HashJoinNode.java]: why you change conjuncts to predicates?
Better to keep it conjuncts, becuase it means `and` and same with `otherJoinConjuncts`

File: be/src/exec/hash_table.h [hunk 6]
```diff
@@ -384,6 +386,12 @@ class HashTable {
 
     // Number of Tuple* in the build tuple row
     const int _num_build_tuples;
+    // the row in hash table is preserved such as RIGHT_OUTER_JOIN
+    const bool _null_preserved;
+    // true: the null value should be keep in 
+    // false: the null value should be filter
```
Comment (imay) [inline be/src/exec/hash_table.h]: what keep in means? and what filter means?

File: fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 7]
```diff
@@ -384,13 +384,13 @@ private PlanFragment createHashJoinFragment(HashJoinNode node, PlanFragment righ
             // TODO: create equivalence classes based on equality predicates
 
             // first, extract join exprs
-            List<Pair<Expr, Expr>> eqJoinConjuncts = node.getEqJoinConjuncts();
+            List<BinaryPredicate> eqJoinPredicates = node.getEqJoinConjuncts();
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java]: name?

**Review by morningman — DISMISSED** [review]
lgtm

**Review by morningman — COMMENTED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**